[chip-test] alert_handler_reverse_ping_in_deep_sleep

moidx opened the pull request:
Check that escalation receivers located inside always-on blocks do not
auto-escalate due to the reverse ping feature while the system is in deep
sleep.

Signed-off-by: Miguel Osorio <miguelosorio@google.com>

engdoreis commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,220 @@
+          .phase = kDifAlertHandlerClassStatePhase0,
+          .signal = 0,
+          .duration_cycles =
+              udiv64_slow(kTestParamAlertHandlerPhase0EscalationDurationUsec *
```
Nit. You can use the new `alert_handler_testutils_get_cycles_from_us` function here.

engdoreis commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,220 @@
+  dif_alert_handler_class_config_t class_config = {
+      .auto_lock_accumulation_counter = kDifToggleDisabled,
+      .accumulator_threshold = 0,
+      .irq_deadline_cycles = udiv64_slow(
```
Nit. Same as above.

engdoreis commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,220 @@
+    irq_external_ctrl(true);
+
+    uint32_t wakeup_threshold =
+        udiv64_slow(kTestParamWakeupThresholdUsec * kClockFreqAonHz, 1000000,
```
Nit. You can use `aon_timer_testutils_get_aon_cycles_from_us` here.

engdoreis commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,220 @@
+        &alert_handler, kDifAlertHandlerLocalAlertAlertPingFail, &is_cause));
+    CHECK(!is_cause, "Unexpected local alert cause: alert_ping_fail");
+    return true;
+  } else {
```
Nit. I would drop this `else` and put the code below outside the block, as the previous blocks either returns or fails.

engdoreis commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,220 @@
+      .classes = classes,
+      .class_configs = class_configs,
+      .classes_len = ARRAYSIZE(class_configs),
+      .ping_timeout = udiv64_slow(
```
Nit. Same as above.

engdoreis commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,220 @@
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "alert_handler_regs.h"  // Generated.
```
I thinks that this header can be removed.

engdoreis reviewed (approved):
LGTM, I just left a few minor comments.

cindychip commented on hw/top_earlgrey/data/chip_testplan.hjson:
````diff
@@ -1387,9 +1387,51 @@
+            4  * N_ESC_SEV * (2 * 2 * 2^PING_CNT_DW)
+            ```
+
+            `pwrgr` is the only block consuming the `N_ESC_SEV` and `PING_CNT_DW` compile time
````
```suggestion
            `pwrmgr` is the only block consuming the `N_ESC_SEV` and `PING_CNT_DW` compile time
```

moidx replied:
Done.

cindychip commented on hw/top_earlgrey/data/chip_testplan.hjson:
```diff
@@ -1387,9 +1387,51 @@
+
+            - On POR reset:
+              - Enable all alerts assigning them to ClassA.
+              - Enable local alert `kDifAlertHandlerLocalAlertAlertPingFail` and assign to ClassB.
```
Should we also enable the local alert `kDifAlertHandlerLocalAlertEscalationPingFail`
I thought the reverse ping timeout is for escalation protocols?

moidx replied:
Updated test plan to enable and check all local alerts. I am also checking that there were no interrupts fired.

cindychip commented on hw/top_earlgrey/data/chip_testplan.hjson:
```diff
@@ -1387,9 +1387,51 @@
+              - Ensure reset status is low power exit. A `kDifRstmgrResetInfoEscalation` signals that
+                there was a local escalation and should result in test failure.
+              - Disable AON timer.
+              - Check there are no flagged `kDifAlertHandlerLocalAlertAlertPingFail` alerts.
```
Same comment as above, should that be `kDifAlertHandlerLocalAlertEscalationPingFail`?

moidx replied:
Updated test plan to enable and check all local alerts. I am also checking that there were no interrupts fired.

moidx commented:
Hi @cindychip, I updated the test to enable and check all local alerts. I also added a flag to the ISR to fail the test if there are any alert interrupts detected. PTAL. Thanks!

cindychip commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,229 @@
+    OT_UNREACHABLE();
+  } else if (pwrmgr_testutils_is_wakeup_reason(
+                 &pwrmgr, kDifPwrmgrWakeupRequestSourceFive)) {
+    LOG_INFO("Wakeup reset");
```
Will this generate a RESET? If so, I wonder if it will clear alert_handler's cause register? I think it would still be good to check the alert_cause and loc_alert_cause, but is there any information that will maintain during reset? Maybe `rst_cause`?

cindychip reviewed (commented):
Thanks Miguel. great test!
I am running in DV and trying to dump waves to see if it is what we expect.

moidx commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,229 @@
+    .hart_id = kPlicTarget,
+};
+
+static alert_handler_isr_ctx_t alert_handler_ctx = {
```
It is used in the ISR to ack any interrupts that may fire, but I can remove it. It should not cause any side effects.

cindychip replied:
Sure, I think it would be nice to remove it. Because this one expect to see an interrupt from class B, but in this test, we do not expect to see any interrupt :)

cindychip commented on sw/device/tests/alert_handler_reverse_ping_in_deep_sleep_test.c:
```diff
@@ -0,0 +1,251 @@
+  kTestParamCycleThroughAllPingsUsec = kTestParamWakeupThresholdUsec >> 2,
+  kTestParamAlertHandlerIrqDeadlineUsec = 100,
+  kTestParamAlertHandlerPhase0EscalationDurationUsec = 100,
+  kTestParamAlertHandlerPingTimeoutUsec = 20,
```
I think 20 us is quite a long time, I would suggest something shorter to expose the error faster.
In the register description it says:
It is recommended to set this value to the equivalent of 256 cycles of the slowest alert sender clock domain in the system (or greater).
I think the value 256 * AON clock period should be fine :)

cindychip replied:
Actually I did the conversion wrong, 20us is fine :) Sorry for the noise.

cindychip reviewed (approved):
Looks great, thanks Miguel

msfschaffner commented on hw/top_earlgrey/data/chip_testplan.hjson:
```diff
@@ -1415,9 +1415,53 @@
+              - Ensure reset status is low power exit. A `kDifRstmgrResetInfoEscalation` signals that
+                there was a local escalation and should result in test failure.
+              - Disable AON timer.
+              - Check there are no flagged local alerts.
```
This LGTM!

msfschaffner commented on hw/top_earlgrey/dv/chip_sim_cfg.hjson:
```diff
@@ -778,15 +778,33 @@
       en_run_modes: ["sw_test_mode_test_rom"]
+      // Disable scoreboard to avoid incorrect alert prediction from the alert_monitor. Due to the
+      // cross-domain alert senders and receivers, the monitor from the chip level did not support
+      // processing alerts accurately from both ends.
```
Is this something that we track in an issue to fix later (if not, should it be?)?

msfschaffner reviewed (approved):
LGTM modulo @cindychip's comments.

cindychip commented:
@moidx here is the run result in dv:

Maybe we can set the `run_timeout_min` to 4 hours?

```
### Test Results
|  Stage  |                       Name                       | Tests                                            |  Max Job Runtime  |  Simulated Time  |  Passi
ng  |  Total  |  Pass Rate  |
|:-------:|:------------------------------------------------:|:-------------------------------------------------|:-----------------:|:----------------:|:------
---:|:-------:|:-----------:|
|   V2    | chip_sw_alert_handler_reverse_ping_in_deep_sleep | chip_sw_alert_handler_reverse_ping_in_deep_sleep |      3.443h       |    255.146ms     |     1 
    |    1    |  100.00 %   |
|   V2    |                                                  | **TOTAL**                                        |                   |                  |     1 
    |    1    |  100.00 %   |
|         |                                                  | **TOTAL**                                        |                   |                  |     1 
    |    1    |  100.00 %   |

```

tjaychen commented:
hey all,
i've been staring at this test a bit, and now i'm not sure how we really should be testing this. Because it's IMO a combination of hardware / software / system behavior.

From what I can see, the most likely way a failure will trigger is a combination of two things:

1. we screw up the clock / reset controls somehow such that `pwrmgr` escalation receiver gets out of sync from the `alert_handler`. 
2. after a low power event, the software for whatever reason takes an enormous amount of time before it enables pings again.

The first event will cause the pwrmgr escalation receiver to think that it should keep counting even though the alert handler has gone offline. 
While the latter (if the limit is 175ms..it's not inconceivable software might run into this, since as far as I can tell rom doesn't enable the ping) would mean the alert handler doesn't ping the escalation for some time, causing it to auto-escalate. 

The test case we have here I feel like is not very likely to ever flag an error (though it is still good to have) because most of the "time" in the test I think is used in pre-sleep, and it's not very likely things will go out of sync there. 

The only way I can think of to test this is to hit the alert handler / pwrmgr combintation repeatedly with different kinds of sleep over and over with real software / or introduce artificial delays on the software, but all of this is going to take way too long in simulation.  We could run it in FPGA, but this would be something that should be run for hours. 

I think at a minimum I should probably add some assertions to ensure the escalation receiver and alert_handler cannot get out of sync, but beyond that I don't really know how to make this better.... any thoughts?

tjaychen commented:
i think what i'm trying to say is that the purpose of these tests is probably to look for places where alert handler and pwrmgr get out of sync. But in this test, there's really only one moment where that could happen.  And even if things DID get out of sync, it would require software to behave in an expected way to actually manifest an error.  So I wonder if it makes more sense to codify what we think are the properties into cross module assertions, and then just try to stimulate it as much as possible.

moidx commented:
Hi @tjaychen, for the long running test you are describing, we could simulate different `pwrmgr` transitions in a similar way to [`pwrmgr_deep_sleep_all_reset_reqs_test.c`](https://github.com/lowRISC/opentitan/blob/master/sw/device/tests/sim_dv/pwrmgr_deep_sleep_all_reset_reqs_test.c) with additional random delays in software while enabling/checking alerts on each power cycle. I can probably write that as a separate test if it makes sense. wdyt?

cindychip commented:
Thanks Tim and Miguel for the suggestions.

I think Tim's concern is totally valid, because `prim_esc_receiver` does not have any status register to tell which state is it at, if reset happened or not, or if this counter is reset to 0.

1).  To address `pwrmgr` and `prim_esc_receiver` not in sync: I can only think of probing some internal signals, or write assertions, as Tim suggested.
To unblock this PR from reviewing/merging, I will create a separate issue for that.

2). To address `pwrmgr` and `alert_handler` not in sync: 
@moidx's suggestion would work!

msfschaffner commented:
@tjaychen just to clarify, the out of sync scenario you mention should not be possible if everything is implemented correctly, right? i.e., the vectors you describe would only help to bias the test so that the involved mechanisms are used in a way that a bug (if there was one) would manifest more readily?

tjaychen commented:
> @tjaychen just to clarify, the out of sync scenario you mention should not be possible if everything is implemented correctly, right? i.e., the vectors you describe would only help to bias the test so that the involved mechanisms are used in a way that a bug (if there was one) would manifest more readily?

yeah. But I worry I'm being too specific on just clk/rst, so if you can think of any other areas we should target..we should probably think about that too.
